pkg/controllers/nodeclaim/underutilization/controller.go

```diff
@@ -0,0 +1,450 @@
+	return "nodeclaim.underutilization"
+}
+
+// IsConsolidated returns true if nothing has changed since markConsolidated was called.
+func (c *Controller) IsConsolidated() bool {
+	return c.lastConsolidationState.Equal(c.cluster.ConsolidationState())
+}
+
+// markConsolidated records the current state of the cluster.
+func (c *Controller) markConsolidated() {
+	c.lastConsolidationState = c.cluster.ConsolidationState()
+}
+
+// Reconcile will try to find consolidation decisions in the cluster:
+//  1. First get the active consolidation decisions in the cluster and see if any decisions
+//     can be invalidated, removing the Underutilization condition from the nodeclaims if invalid.
+//  2. If there are no actively progressing consolidations, do not do anything until cluster
+//     state is no longer in a consolidated state. This is important to reduce resource usage
+//     when the cluster is idle.
+//  3. If there are 5 or more actively progressing consolidations, exit early, as we should just
+//     wait for these to finish.
+//  4. Compute a consolidation decision with scheduling simulations
+//  5. If we find a valid decision, mark all node claims as Underutilized.
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	noms := c.getActiveConsolidations(ctx)
+	// If there are no active nominations and it's been less than 5 minutes since the last time
+	// the cluster was marked as consolidated, wait.
+	if len(noms) == 0 && c.IsConsolidated() {
+		return reconcile.Result{RequeueAfter: time.Until(c.cluster.ConsolidationState().Add(5 * time.Minute))}, nil
+	}
+	// If there are too many nominations, see if we can quickly invalidate any existing nominations
+	if len(noms) >= maxConcurrentNominations {
+		for _, cmd := range noms {
+			// If any nomination is invalid, remove the nomination and requeue
+			// Short circuit and only unmark one of the commands to optimistically save work done
+			if err := c.validateCandidates(ctx, cmd.Candidates...); err != nil {
+				err = multierr.Append(err, c.setNodeClaimsUnderutilized(ctx, false, "", cmd.Candidates...))
+				return reconcile.Result{}, fmt.Errorf("validating candidates, %w", err)
+			}
+		}
+		// If we can't invalidate any consolidations, we shouldn't enqueue another. Try again later.
+		return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
+	}
+
+	newCandidateSet, err := c.computeConsolidation(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("computing consolidation decision, %w", err)
+	}
+	// we failed to find a consolidation decision
+	if len(newCandidateSet) == 0 {
+		c.markConsolidated()
+		return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
+	}
+
+	if err := c.setNodeClaimsUnderutilized(ctx, true, uuid.NewUUID(), newCandidateSet...); err != nil {
+		return reconcile.Result{}, fmt.Errorf("propagating nomination, %w", err)
+	}
+
+	return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
```

Comment:
There's no comment next to this but I'm assuming that this is a heuristic to try to reduce the number of times that we are running this async?